staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -513,3 +515,89 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
 func (u unionTransformers) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) (out []byte, err error) {
 	return u[0].TransformToStorage(ctx, data, dataCtx)
 }
+
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformerOverrides *atomic.Value
+	closeTransformers    *atomic.Value
+	inFlight             *atomic.Value
+}
+
+// NewDynamicTransformers returns a new DynamicTransformers.
+func NewDynamicTransformers(transformerOverrides map[schema.GroupResource]value.Transformer, closeTransformers context.CancelFunc) *DynamicTransformers {
+	dynamicTransformers := &DynamicTransformers{
+		transformerOverrides: &atomic.Value{},
+		closeTransformers:    &atomic.Value{},
+		inFlight:             &atomic.Value{},
+	}
+
+	// new inFlight variable to track transformers that will be set in here.
+	var inFlight atomic.Int32
+	// ignore the return results here since we do not want to close the initial transformers.
+	// return results are only used in hot reload controller to close the old transformers after the new ones are loaded.
+	dynamicTransformers.Set(transformerOverrides, closeTransformers, &inFlight)
+
+	return dynamicTransformers
+}
+
+// TransformerForResource returns the transformer for the given resource.
+func (d *DynamicTransformers) TransformerForResource(resource schema.GroupResource) value.Transformer {
+	return &resourceTransformer{
+		resource:             resource,
+		transformerOverrides: d.transformerOverrides,
+		inFlight:             d.inFlight,
+	}
+}
+
+// Set sets the transformer overrides.
+func (d *DynamicTransformers) Set(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+	inFlight *atomic.Int32,
+) (context.CancelFunc, *atomic.Int32) {
+	oldCloseTransformers := d.closeTransformers.Load()
+	oldInFlight := d.inFlight.Load()
+
+	// store new values
+	d.transformerOverrides.Store(transformerOverrides)
+	d.closeTransformers.Store(closeTransformers)
+	d.inFlight.Store(inFlight)
+
+	if oldCloseTransformers != nil && oldInFlight != nil {
```

Comment:
There should be no `nil` checks - this should always work.

Reply:
No.  For the first time, they will be nil. Same explanation as https://github.com/kubernetes/kubernetes/pull/112050#discussion_r999969690

Reply:
removed nil checks